File: 0085-maximal-rectangle/0085-maximal-rectangle.cpp
class Solution {
public:
   int helper(vector<int>&arr){
  int ans =0,idx=0; int n= arr.size();
  stack<int>st;
  for(int i=0 ; i< arr.size();i++){
     while(!st.empty() && arr[st.top()]>arr[i]){
         int idx = st.top();
         st.pop();
         if(!st.empty())
         ans = max(ans,arr[idx] * (i-st.top()-1));
         else ans = max(ans,arr[idx]*i);
     }

     st.push(i);
  }

  while(!st.empty()){
      int idx = st.top();
      st.pop();
      if(!st.empty()){
      ans = max(ans,arr[idx]*(n-st.top()-1));}
      else { ans = max(ans,(arr[idx]*n));}
  }
  return ans;
}
    int maximalRectangle(vector<vector<char>>& matrix) {
        vector<int>arr(matrix[0].size(),0); int maxi_wid = 0;
        int row = matrix.size(); int col = matrix[0].size();
        for(int j=0; j<row; j++){
        for(int i=0; i<col; i++){
            if(matrix[j][i] == '0'){
                arr[i] = 0;
            }else{
                arr[i] = arr[i]+1;
            }
        }
        maxi_wid = max(maxi_wid,helper(arr));
        }
        return maxi_wid;
    }
};

File: 0085-maximal-rectangle/0085-maximal-rectangle.java
class Solution {
    int FindArea(int[] height){
        int ans =0;
        Stack<Integer>st = new Stack<>();
      for(int i=0;i<height.length;i++){
         while(!st.empty() && height[st.peek()] >= height[i]){ 
         int curr = st.peek(); st.pop();
         int prev = (st.empty())?-1:st.peek();
          int len = i - prev -1;
         int area = len * height[curr];
         ans = Math.max(ans,area);
         }
         st.push(i);


      }

      while(!st.empty()){
       int curr = st.peek(); st.pop();
         int prev = (st.empty())?-1:st.peek();
         
         
          int len = height.length - prev -1;
         int area = len * height[curr];
         ans = Math.max(ans,area);
      }

      return ans;
    }
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) return 0;
        int ans=0;
        int[]height = new int[matrix[0].length];
        Arrays.fill(height,0);
        for(int i=0;i<matrix.length;i++){
            for(int j=0;j<matrix[0].length;j++){
                 if(matrix[i][j] == '1'){
                    height[j]++;
                 }
                 else{
                    height[j] = 0;
                 }
            }
            int area = FindArea(height);
            ans = Math.max(ans,area);
        }
        return ans;
    }
}